test/integration/good/code-gen/cl.expected

```diff
@@ -146,7 +163,9 @@ class optimize_glm_model : public model_base_crtp<optimize_glm_model> {
   std::vector<int> y_vi_d;
   std::vector<int> y2_vi_d;
   int y_s_d;
+  double y_r_d;
   matrix_cl<double> X_d_opencl__;
+  matrix_cl<double> y_r_d_opencl__;
```

Comment:
interesting, maybe this is the wrong place for this question but do we really need to create a matrix to hold a single double? Can't our OpenCL kernels just take a double as one of the arguments? I'm guessing the answer is "yes we could do that but then we'd need to overload all of the kernels to do that, and that's hard for X reason" is that right?

Reply:
You are right. 

You either supply the kernel a cl::Buffer or a scalar. The cl::Buffer lives in the GPU memory (values need to be copied there beforehand) while the scalar is passed by value with the kernel call and the scalars lives in the CPU  memory. So its not only an overload, but you need to handle that variable completely different in this case. You also cant modify the value tho that is not a problem in this case. So that is the reason we are handling it this way.

Its far from optimal because we hold an entire object for a single value. If anyone has any suggestions on how we could improve this, they are more than welcome.

Reply:
> and the scalars lives in the CPU memory. 

Ahh, wait, does this mean the kernels would just have to send the double over to the GPU anyway separately? I was imagining the "store the double in its own cl::Buffer" approach would lead to increased latency because I thought that the double could have been sent over to the GPU *as part of* the "command" to run the kernel. If there's no performance or latency difference then this cl::Buffer approach seems optimal to me as you get more code re-use.

Reply:
No no, you are right. In this case it would get sent over as part of the command. That is why I said it lives in the CPU memory. Sorry for any confusion. 

We use this for kernels where we know an argument is going to always be a read-only scalar. For instance in `fill()` `val` is sent over as part of the "execute kernel" command:

`__kernel void fill(__global double* A, double val,`

Lets look at another example with rep_matrix, where B can a scalar, vector or row_vector. Here we always use cl::buffers and by extent matrix_cl which is essentially an object that holds the cl::buffer:
`__kernel void rep_matrix(__global double* A, __global double* B,`

If we wanted to optimize the case where B is a scalar to this:
`__kernel void rep_matrix(__global double* A, double B,`

we would also need to have a copy of the entire kernel since its no longer using pointers for B. Also in the case of where the scalar resides on the GPU already we would still need to use the first version. It would be slightly more efficient for this case tho.

Reply:
Gotcha. So the tradeoff is basically creating duplicates of the kernels just to change the function signature, and in this case all that buys you is one less to_opencl copy command during model construction, i.e. just once at the beginning of the program, and with the async stuff probably not adding any latency at all given it's also sending a much larger matrix over at the same time, right?

So probably not worth changing anything, right? Thanks for the discussion though!

Reply:
Yes, that is basically it. Return on investment is not worth it :) No problem, happy to discuss.